controllers/datamovement_controller.go

```diff
@@ -214,40 +254,154 @@ func (r *DataMovementReconciler) Reconcile(ctx context.Context, req ctrl.Request
+		var combinedOutBuf, parseBuf bytes.Buffer
+		cmd.Stdout = io.MultiWriter(os.Stdout, &parseBuf, &combinedOutBuf)
+		cmd.Stderr = cmd.Stdout // Combine stderr/stdout
+
+		// Use channels to sync progress collection and cmd.Wait().
+		chCommandDone := make(chan bool, 1)
+		chProgressDone := make(chan bool)
+
+		// Start the data movement command
+		cmd.Start()
+
+		// While the command is running, capture and process the output. Read
+		// lines until EOF to ensure we have the latest output. Then use the
+		// last regex match to obtain the most recent progress.
+		go func() {
+			var progressInt32 *int32
+			var elapsed metav1.Duration
+			elapsed.Duration = 0
+			progressStart := metav1.NowMicro()
+
+			// Perform the actual collection and update logic
+			parseAndUpdateProgress := func() {
+				progressOutput := ""
+
+				// Read all lines of output until EOF
+				for {
+					line, err := parseBuf.ReadString('\n')
+					if err == io.EOF {
+						break
+					} else if err != nil {
+						log.Error(err, "failed to read progress output")
+					}
+					progressOutput += line
+				}
+
+				// Get the times now before we start the parsing
+				progressNow := metav1.NowMicro()
+				elapsed.Duration = progressNow.Time.Sub(progressStart.Time)
+
+				// Find all matching progress matches but we only care about the
+				// last (most recent one)
+				groups := progressRe.FindAllStringSubmatch(progressOutput, -1)
+				if len(groups) > 0 {
+					mostRecent := groups[len(groups)-1]
+
+					// Get the progress and convert to int
+					progress, err := strconv.Atoi(mostRecent[3])
+					if err == nil {
+						// Initialize the pointer if we haven't done so already.
+						// We don't want to do this above since we do not want an
+						// initial value of 0 unless we actually have a progress
+						if progressInt32 == nil {
+							progressInt32 = new(int32)
+						}
+						*progressInt32 = int32(progress)
+					} else {
+						log.Error(err, "failed to parse progress output", "progressMatch", mostRecent)
+					}
+				}
+
+				// Update the CommandStatus in the DM resource
+				err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+					dm := &nnfv1alpha1.NnfDataMovement{}
+					if err := r.Get(ctx, req.NamespacedName, dm); err != nil {
+						return client.IgnoreNotFound(err)
+					}
+
+					// Update progress if we have one
+					if progressInt32 != nil {
+						cmdStatus.ProgressPercentage = progressInt32
+					}
+					// Update the output and times regardless
+					cmdStatus.LastMessage = progressOutput
+					cmdStatus.LastMessageTime = progressNow
+					cmdStatus.ElapsedTime = elapsed
+
+					if dm.Status.CommandStatus == nil {
+						dm.Status.CommandStatus = &nnfv1alpha1.NnfDataMovementCommandStatus{}
+					}
+					cmdStatus.DeepCopyInto(dm.Status.CommandStatus)
+
+					log.Info("Updating Progress", "CommandStatus", dm.Status.CommandStatus)
+					return r.Status().Update(ctx, dm)
```

Comment:
I think this would cause unnecessary updates; you should trigger updates only if the LastMessage changes from the previous update. Right now you're forcing updates every second, but there are parts of dcp that aren't affected by the --progress flag (setup and teardown); and we may find that --progress [time] slows down the program so we move to `--progress 30` or something.

Reply:
`dcp` is outputting the progress every second (via `--progress 1`), but we're only collecting that output at an interval of 5 seconds (configurable via the config map). You make a good point though. I orginally had a check to see if `progressOutput != ""`, which I think will work here. `progressOutput` will always be different than `lastMessage` unless there is nothing new to read (i.e. ""). As I mentioned above, `lastMessage` will not be the whole buffer - just the new lines output.

Reply:
I meant for `lastMessage` to be the last line (singular) read from output, not the last lines (plural) since last capture.

The idea was if the below happened
```
t=0 0% <- trigger
t=1 1%
t=2 2%
t=3 3%
t=4 4%
t=5 5% <- trigger
``` 
Then at t=0 you would see the status `lastMessage: "t=0 0%"`
And the next reading after t=5 would see the status `lastMessage: "t=5 5%"`

You don't need to capture the t=1 through t=4 messages since that is safely assumed.

Reply:
How it works now is that at t=5, you'll get the following for `lastMessage`: 
```
t=1 1%
t=2 2%
t=3 3%
t=4 4%
t=5 5%
```
Then the regex comes through and matches the whole thing and it will only use the last match (`t=5`) and use that for the progress. The reasoning behind this is that when the dcp command finishes, we'll get more output after the 100%/done line so we need to make sure we're capturing everything.

But, to address your concern, I think we can just use that last match (`t=5 5%`) and set that as `lastMessage` rather than the whole thing. Does that address your concern?

Reply:
Ohh I see why you did it that way now. Instead of accumulating the lines, run the percentage parser per line. Then make `lastMessage` the last line you read.

Reply:
Another explainer - the reason I wanted CommandStatus to be brief and not comprehensive is that this should also apply to the Copy-Offload API - particularly the Status response. In trying to keep the GRPC messages slim, its one `lastMessage` line (while we can't guarantee the string is short, it is coming from a command-line program, so it's likely < 80 characters by convention)

Reply:
I will change this so that we're only getting the last matched regex to keep it short.